### PR TITLE
Split things up and pare them down

### PR DESCRIPTION
**Pull request by tfausak**

### CODE REVIEW

File: src/test/Main.hs [hunk 1]
```diff
@@ -1,162 +1,428 @@
...
+  , unsafeInto @Int8 (1 :: Int16) ~?= 1
+
+# if MIN_VERSION_GLASGOW_HASKELL(8, 10, 1, 0)
+  , ($$( liftedCast (1 :: Int16) ) :: Int8) ~?= 1
+  , ($$( liftedFrom @Int16 1 ) :: Int8) ~?= 1
+  , ($$( liftedInto @Int8 (1 :: Int16) ) :: Int8) ~?= 1
+# else
+  , $( liftedCast @Int16 @Int8 1 ) ~?= (1 :: Int8)
+  , $( liftedFrom @Int16 @Int8 1 ) ~?= (1 :: Int8)
+  , $( liftedInto @Int8 @Int16 1 ) ~?= (1 :: Int8)
```
Comment (tfausak) [inline src/test/Main.hs:36]: The ergonomics of these untyped Template Haskell splices are pretty bad. They need to be fully specified because type inference doesn't go "into" or "from" the splice. For example, this is ambiguous:

``` hs
-- ambiguous
$( liftedFrom @Int16 1 ) :: Int8
-- needs to be
$( liftedFrom @Int16 @Int8 1 ) :: Int8
```

Even though the result must be an `Int8`, that information doesn't make its way to the inside of the splice. Similar, this is ambiguous:

``` hs
-- ambiguous
$( liftedCast @Int16 @Int8 1 ) ~?= 1
-- needs to be
$( liftedCast @Int16 @Int8 1 ) ~?= (1 :: Int8)
```

Even though the result has to be an `Int8`, that information doesn't make its way out of the splice.

**Comment by tfausak** [comment]
I still need to write instances between all the various signed and unsigned types. I have `Int8` to `Int16` and `Word8` to `Word16`, but not `Int8` to `Word8`. 

There's a huge number of instances to provide. Should I write some Template Haskell functions to generate all the declarations, or just do it by hand? Doing it by hand isn't very hard. 

Also I'm assuming that `Int` is never bigger than `Int64`, but I don't know if that's always true. Are there any 128-bit architectures? I'm assuming either (a) no or (b) there may be but they're not supported by GHC. Case in point, there's not 128-bit `Int128` type.

File: src/test/Main.hs (hunk 1 above)
Comment (tfausak) [reply src/test/Main.hs]: Ah, I didn't realize that GHC 8.8 supported typed Template Haskell splices like `$$(...)` even though it doesn't have `liftTyped`. Should be fixed now. 

This also makes documentation easier: Everything uses `$$(...)` regardless of GHC version.

**Comment by tfausak** [comment]
Very strange that one of the tests failed on GHC 9.0.1:

```
### Failure in: Witch:81:TryCast Integer Natural:1
src/test/Main.hs:443
expected: Right 18446744073709551616
 but got: Right 0
```

This is the test case:

``` hs
tryCast @Integer @Natural 18446744073709551616 ~?= Right 18446744073709551616
```

That should work fine. The number intentionally overflows the range of `Int` and `Int64`, but that shouldn't matter because it's an `Integer`. And indeed the test passes with other versions of GHC. Is this a GHC bug?

**Comment by tfausak** [comment]
What I've got in this PR so far covers the majority of the common numeric types, but there's still a few more: `Float`, `Double`, `Ratio`, `Fixed`, and `Complex`. (`Complex` is questionable. I've never seen anyone use it. Then again, I've never had reason to use it.) Most of the conversions are straightforward, but I do have two things to figure out:

- When is it safe to convert between `Float` and `Double`? I think it's safe to assume that any `Float` can be represented as a `Double`, but going the other way feels tricky. I may need to play around a little with some property tests to figure out what to do here. 
- Is there an easy way to figure out if a floating point number represents an integer value exactly? I think the clunky way is to make sure that `floor` and `ceiling` return the same value.

**Comment by tfausak** [comment]
I really have no idea what's going on here.

``` hs
ghci> fromIntegral (18446744073709551616 :: Integer) :: Natural
18446744073709551616

ghci> fromIntegral @Integer @Natural 18446744073709551616
18446744073709551616

ghci> f x = if x < 0 then Nothing else Just $ fromIntegral x
ghci> f (18446744073709551616 :: Integer) :: Maybe Natural
Just 18446744073709551616
```

But it's definitely failing on all the GHC 9.0.1 builds.

**Comment by tfausak** [comment]
Ah, it's this bug: https://mail.haskell.org/pipermail/haskell-cafe/2021-March/133540.html

**Comment by tfausak** [comment]
I'm struggling to wrap my head around what it even means to "safely" convert from or into floating point types. Take `tryCast @Float @Rational` for example:

- Should `-0` succeed and become `0` or should it fail?
- Clearly all of `0 / 0`, `1 / 0`, and `-1 / 0` should fail. 
- What should happen around `16777216`? That's the largest single precision floating point value that can precisely represent an integer. In other words, `16777217` doesn't actually exist but `16777218` does. Should converting `16777218` fail? If so, why? If not, why not? 
  - Arguably it should succeed since the input value can be represented exactly in the output. And what makes this value special? The range of `Float` values is inherently lossy, so why does it matter than `16777217` doesn't exist? 
  - On the other hand, arguably it should fail because if you want to convert _from_ an `Int` value to a `Float` value, once you get above `16777216` all bets are off. I don't think this argument is as compelling.

**Comment by tfausak** [comment]
Alright, I think I've made some reasonable decisions. I still have some questions though:

- Should converting from `Double` to `Float` be partial?
- How can I determine if it's safe to convert from `Rational` to `Float` (or `Double`)? 
- Does it make sense for `Rational` to accept big integral values but the various integral types to reject them?

**Comment by tfausak** [comment]
> Should -0 succeed and become 0 or should it fail?

It should succeed, because `0 == (-0 :: Float)`. 

> What should happen around 16777216?

When dealing with integral types, the range of acceptable float values should be ±16777215. Values outside that range aren't exactly representable, so you don't know if 16777216 is really that or was supposed to be 16777217.

> Should converting from Double to Float be partial?

I'm still grappling with this. On the one hand, it's nice for equality to distribute over `cast`. That's what the property tests are checking. On the other hand, both representations are already lossy so it's not like converting from `Int64` to `Int32`. 

> How can I determine if it's safe to convert from Rational to Float (or Double)?

Basically the same answer as the previous question. Rational values are arbitrary precision, so you're probably almost always going to lose something by going into floats or doubles. 

> Does it make sense for Rational to accept big integral values but the various integral types to reject them?

Yes. There's not really anything special about the max integral value for rationals. What about the max value with 0.1 precision? Or 0.01?

**Comment by tfausak** [comment]
I don't think the property tests are very useful. I added `instance Cast Double Float where cast = realToFrac` and even with 16,000 successes it didn't find a counter example. That makes sense though: What are the odds that it arbitrarily checks two `Double` values that get converted into the same `Float` value? Basically zero. 

Since I'm not sure what to do, I'm going to avoid providing any instances for now.

**Comment by tfausak** [comment]
For some reason it's particularly difficult for me to wrap my head around floats. So let me consider a different instance: converting between lists and sets. Clearly different lists can become the same set: both `[ 1 ]` and `[ 1, 1 ]` become `{ 1 }`. 

Does that mean it should be a `TryCast` instance that checks for duplicates? I feel like that would be annoying for little benefit. 

Alternatively I could only allow converting into a multi-set (or something like `Map k Natural`). But then of course you'll want to convert that multi-set into a regular set, and you have the same problem again. Both `{ 'a' => 1 }` and `{ 'a' => 2 }` would become `{ 'a' }`.

The crux of the problem here is that I don't have any laws to guide which instances to provide. I think injectivity is desirable, but also it appears to limit useful instances. If I ditch the laws and implement what "makes sense" to me, then I'll probably wind up doing something stupid sooner or later. 

This all could suggest that I need another type class for known lossy conversions. Something that communicates it's possible to lose precision, but that's also probably what you want. Unfortunately I think the semantics of that type class would be even harder to pin down. For example if you want to do a lossy conversion from `Int32` to `Word32`, do you clamp the output or preserve the representation? Both could make sense.
